drivers/timer/nrf_grtc_timer.c

```diff
@@ -0,0 +1,571 @@
+	}
+#else
+	nrfy_grtc_sys_counter_active_state_request_set(NRF_GRTC, true);
+	k_busy_wait(GRTC_SYSCOUNTERVALID_SETTLE_MAX_TIME_US);
+#endif
+}
+
+static inline void grtc_wakeup(void)
+{
+	if (IS_ENABLED(CONFIG_NRF_GRTC_SLEEP_ALLOWED)) {
+		grtc_active_set();
+	}
+}
+
+static inline void grtc_sleep(void)
+{
+	if (IS_ENABLED(CONFIG_NRF_GRTC_SLEEP_ALLOWED)) {
+#if defined(NRF_GRTC_HAS_SYSCOUNTER_ARRAY) && (NRF_GRTC_HAS_SYSCOUNTER_ARRAY == 1)
+		nrfy_grtc_sys_counter_active_set(NRF_GRTC, false);
+#else
+		nrfy_grtc_sys_counter_active_state_request_set(NRF_GRTC, false);
+#endif
+	}
+}
+
+static inline uint64_t counter_sub(uint64_t a, uint64_t b)
+{
+	return (a - b);
+}
+
+static inline uint64_t counter(void)
+{
+	uint64_t now;
+
+	grtc_wakeup();
+	nrfx_grtc_syscounter_get(&now);
+	grtc_sleep();
+	return now;
+}
+
+static inline uint64_t get_comparator(uint32_t chan)
+{
+	uint64_t cc;
+	nrfx_err_t result;
+
+	result = nrfx_grtc_syscounter_cc_value_read(chan, &cc);
+	if (result != NRFX_SUCCESS) {
+		if (result != NRFX_ERROR_INVALID_PARAM) {
+			return -EAGAIN;
+		}
+		return -EPERM;
+	}
+	return cc;
+}
+
+static void system_timeout_set(uint64_t value)
+{
+	if (value <= NRF_GRTC_SYSCOUNTER_CCADD_MASK) {
+		grtc_wakeup();
+		nrfx_grtc_syscounter_cc_relative_set(&system_clock_channel_data, value, true,
+						     NRFX_GRTC_CC_RELATIVE_SYSCOUNTER);
+		grtc_sleep();
+	} else {
+		nrfx_grtc_syscounter_cc_absolute_set(&system_clock_channel_data, value + counter(),
+						     true);
+	}
+}
+
+static bool compare_int_lock(int32_t chan)
+{
+	atomic_val_t prev = atomic_and(&int_mask, ~BIT(chan));
+
+	nrfx_grtc_syscounter_cc_int_disable(chan);
+
+	return prev & BIT(chan);
+}
+
+static void compare_int_unlock(int32_t chan, bool key)
+{
+	if (key) {
+		atomic_or(&int_mask, BIT(chan));
+		nrfx_grtc_syscounter_cc_int_enable(chan);
+	}
+}
+
+static void sys_clock_timeout_handler(int32_t id, uint64_t cc_val, void *p_context)
+{
+	ARG_UNUSED(id);
+	ARG_UNUSED(p_context);
+	uint64_t dticks;
+	uint64_t now = counter();
+
+	if (unlikely(now < cc_val)) {
+		return;
+	}
+	if (!IS_ENABLED(CONFIG_TICKLESS_KERNEL)) {
+		/* protection is not needed because we are in the GRTC interrupt
+		 * so it won't get preempted by the interrupt.
+		 */
+		system_timeout_set(CYC_PER_TICK);
+	}
+
+	dticks = counter_sub(now, last_count) / CYC_PER_TICK;
+
+	last_count += dticks * CYC_PER_TICK;
+	sys_clock_announce(IS_ENABLED(CONFIG_TICKLESS_KERNEL) ? (int32_t)dticks : (dticks > 0));
+}
+
+int32_t z_nrf_grtc_timer_chan_alloc(void)
+{
+	uint8_t chan;
+	nrfx_err_t err_code;
+
+	/* Prevent allocating all available channels - one must be left for system purposes. */
+	if (ext_channels_allocated >= EXT_CHAN_COUNT) {
+		return -ENOMEM;
+	}
+	err_code = nrfx_grtc_channel_alloc(&chan);
+	if (err_code != NRFX_SUCCESS) {
+		return -ENOMEM;
+	}
+	ext_channels_allocated++;
+	return (int32_t)chan;
+}
+
+void z_nrf_grtc_timer_chan_free(int32_t chan)
+{
+	IS_CHANNEL_ALLOWED_ASSERT(chan);
```

Comment:
why this check is not done when allocating channel in the function above?
btw - this likely should not be assert but a proper argument check as this function is not internal

Reply:
CC @adamkondraciuk